jwt/jwt.go

```diff
@@ -227,43 +293,75 @@ func UserDetailsFromJWT(ctx context.Context) (firstName, lastName, email string,
 //        }
 //      ]
 //    },
-//    "issued_at": "2020-12-09T16:09:35.004042Z"
 //  },
-//  "sub": "695725d4-a345-4e68-919a-7395cb68484c"
-//}
-func ValidateToken(logger *logrus.Logger, t string) (*jwt.Token, error) {
-	keyForTokenWithLogger := func(token *jwt.Token) (interface{}, error) {
-		return keyForToken(logger, token)
-	}
-	token, err := jwt.Parse(t, keyForTokenWithLogger)
+// }
+func ValidateToken(logger *logrus.Logger, t string) (jwt.Token, error) {
+	_, err := accountsKeySet(logger)
 	if err != nil {
-		return nil, err
+		return nil, errors.AddContext(err, "failed to fetch JWT key set")
 	}
-	if !token.Valid {
-		return nil, errors.New("token is invalid")
+	token, err := jwt.Parse([]byte(t), jwt.WithKeySet(accountsPublicJWKS))
+	if err == nil {
+		// return nil, err
+		return token, nil
 	}
-	// TODO Verify issuer, scope, etc.?
-	return token, nil
-}
 
-// keyForToken finds a suitable key for validating the
-// given token among the public keys provided by Oathkeeper.
-func keyForToken(logger *logrus.Logger, token *jwt.Token) (interface{}, error) {
-	if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
-		return nil, errors.New(fmt.Sprintf("unexpected signing method: %v", token.Header["alg"]))
-	}
+	// try to parse the token as an oathkeeper token
 	keySet, err := oathkeeperPublicKeys(logger)
 	if err != nil {
-		return nil, err
+		return nil, errors.AddContext(err, "failed to fetch Oathkeeper's JWKS")
 	}
-	if reflect.ValueOf(token.Header["kid"]).Kind() != reflect.String {
-		return nil, errors.New("invalid jwk header - the kid field is not a string")
+	token, err = jwt.Parse([]byte(t), jwt.WithKeySet(keySet))
+	if err != nil {
+		return nil, err
 	}
-	keys := keySet.LookupKeyID(token.Header["kid"].(string))
-	if len(keys) == 0 {
-		return nil, errors.New("no suitable keys found")
+	return token, nil
+}
+
+// accountsKeySet checks whether we have the needed key set cached and if we
+// don't it fetches it and caches it for us. The method returns the first key in
+// the set or ann error.
+//
+// Note that this key set includes both the public *and* private keys.
+//
+// See https://tools.ietf.org/html/rfc7517
+// See https://auth0.com/blog/navigating-rs256-and-jwks/
+// See http://self-issued.info/docs/draft-ietf-oauth-json-web-token.html
+// Encoding RSA pub key: https://play.golang.org/p/mLpOxS-5Fy
+func accountsKeySet(logger *logrus.Logger) (jwk.Set, error) {
+	if accountsJWKS == nil {
+		var b []byte
+		var err error
+		switch build.Release {
+		case "dev":
+			b, err = ioutil.ReadFile("jwks.json")
+		case "testing":
+			b, err = ioutil.ReadFile("fixtures/jwks.json")
+		default:
+			b, err = ioutil.ReadFile(AccountsJWKSFile)
+		}
+		if err != nil {
+			logger.Warningln("ERROR while reading accounts JWKS", err)
+			return nil, err
+		}
+		set := jwk.NewSet()
```

Comment:
Is this necessary? Does 
```go
var set jwk.Set
err = json.Unmarshal(b, &set)
```
not work?

Reply:
I couldn't make it work. The `jwk.NewSet()` is this:
```go
func NewSet() Set {
	return &set{}
}
```
So, while the `Set` interface is public, the returned type is unexported. There might be a way to do this without this function call but I'm not sure how.

Reply:
ohhh it's an interface. nevermind then